src/opera/parser/parser.py

```diff
@@ -0,0 +1,159 @@
+        # pylint: disable=protected-access
+        tosca_document_parsed = parser._parse_csar_file(csar, ToscaPath(service_template,
+                                                                        Location(service_template.name, 0, 0)))
+        result = parser.stdlib
+        logger.debug("Merging parsed document and the TOSCA standard library.")
+        result.merge(tosca_document_parsed)
+        logger.debug("Resolving node references.")
+        result.visit("resolve_reference", result)
+
+        return result
+
+    def __init__(self, tosca_version: str, document_type: Type[TDocument]):
+        self.tosca_version = tosca_version
+        self.document_type = document_type
+        self.stdlib: TDocument = self._load_stdlib(self.tosca_version)
+
+    def _parse_csar_file(self, csar: ToscaCsar, tosca_path: ToscaPath) -> TDocument:
+        """
+        Parse a file contained in a CSAR.
+        :param csar: the CSAR file.
+        :param tosca_path: the TOSCA path definition. relative to the CSAR root.
+                           If not present, the CSAR service definition file will be used.
+        """
+        path: pathlib.PurePath = tosca_path.data
+        if path.is_absolute():
+            return self._parse_bare_file(tosca_path)
+        logger.debug("Parsing CSAR file %s.", path)
+
+        with csar.open_member(path) as input_fd:
+            tosca_document_yaml = yaml.load(input_fd, path.name)
+
+        if not isinstance(tosca_document_yaml.value, dict):
+            raise ParseError("The top level structure must be a map.", tosca_document_yaml.loc)
+
+        tosca_document_parsed = self.document_type.parse(tosca_document_yaml)
+        logger.debug("Merging imports from CSAR file %s", path)
+        self._merge_imports(csar, tosca_document_parsed)
+
+        return tosca_document_parsed
+
+    def _parse_bare_file(self, tosca_path: ToscaPath) -> TDocument:
+        """
+        Parses a bare file, not contained in the CSAR.
+        :param tosca_path: the TOSCA path reference.
+        """
+        path: pathlib.PurePath = tosca_path.data
+        logger.debug("Parsing bare file %s.", path)
+        if not path.is_absolute():
+            tosca_path.abort("CSAR-external absolute path '{}' is not absolute.".format(path), tosca_path.loc)
+
+        with open(pathlib.Path(path), "r") as f:
+            tosca_document_yaml = yaml.load(f, path.name)
+        tosca_document_parsed = self.document_type.parse(tosca_document_yaml)
+
+        # we don't merge imports because this is a bare file
+        return tosca_document_parsed
+
+    def _load_stdlib(self, tosca_version: str) -> TDocument:
+        stdlib_yaml = stdlib.load(tosca_version)
+        return self.document_type.parse(stdlib_yaml)
+
+    def _merge_imports(self, csar: ToscaCsar, document: TDocument):
+        """
+        Import paths can be absolute or relative.
+          - absolute paths can be anywhere on the disk
```

Comment:
With CSAR support, we shouldn't do this anymore. Everything should be on the disk and within the CSAR root directory. Nothing outside it should exist for the merge or the artifacts.

Reply:
Great. Do you want me to fix this in this PR or make a new one? It's quite a significant change which would benefit from a separate review process.

This is another divergence from the TOSCA standard, right? There are instances of absolute paths e.g. in [Simple Profile 1.3](https://docs.oasis-open.org/tosca/TOSCA-Simple-Profile-YAML/v1.3/TOSCA-Simple-Profile-YAML-v1.3.html) in Section 4.8.1.3.1.

Reply:
Ugh, so I can craft a CSAR that imports something off the file system that's not within the CSAR? I don't know, it feels like there should be a switch for this kind of icky compliance.

In any case, I'd make it a new PR, but let's discuss it with @tadeboro first.